odoo/addons/base/tests/test_graph.py

```diff
@@ -0,0 +1,363 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import time
+import logging
+from unittest.mock import patch
+from typing import Iterable, List
+
+from odoo.tests.common import BaseCase, TransactionCase
+from odoo.modules.graph import Graph
+from odoo.modules.module import _DEFAULT_MANIFEST
+from odoo.tools import mute_logger
+
+_logger = logging.getLogger(__name__)
+
+class TestGraph(BaseCase):
+    def _test_graph_order(self, manifests: dict, modules_list: Iterable[Iterable[str]], expected: List[str]) -> None:
+        """
+        Test the order of the modules that need to be loaded
+
+        :param manifests: A dictionary of module manifests: key 'depends' is required
+        :param modules_list: [['module_a', 'module_b'], ['module_c'], ...]
+            module_a and module_b will be added in the first round
+            module_c will be added in the second round
+            ...
+        :param expected: expected graph order
+        """
+        manifests = {
+            name: {**_DEFAULT_MANIFEST.copy(), **manifest}
+            for name, manifest in manifests.items()
+        }
+        with patch('odoo.modules.graph.Graph.update_from_db'),\
+                patch('odoo.modules.module.get_manifest', manifests.get):
+            graph = Graph()
+            graph_legacy = GraphLegacy()
+
+            for modules in modules_list:
+                graph.add_modules(None, modules)
+                graph_legacy.add_modules(None, modules)
+
+            names = list(g.name for g in graph)
+            names_legacy = list(g.name for g in graph_legacy)
+
+            self.assertListEqual(names, expected)
+            self.assertListEqual(names_legacy, expected)
+
+    def test_graph_order_1(self):
+        manifests = {
+            'base': {
+                'depends': []
+            },
+            'module1': {
+                'depends': ['base']
+            },
+            'module2': {
+                'depends': ['module1']
+            },
+            'module3': {
+                'depends': ['module1']
+            },
+            'module4': {
+                'depends': ['module2', 'module3']
+            },
+            'module5': {
+                'depends': ['module2', 'module4']
+            },
+        }
+        # modules are in random order
+        self._test_graph_order(
+            manifests,
+            [['base'], ['module3', 'module4', 'module1', 'module5', 'module2']],
+            ['base', 'module1', 'module2', 'module3', 'module4', 'module5']
+        )
```

Comment:
Given the fact that the graph only relies on `depends`, you could pass the modules and their dependencies as a `dict`:
```suggestion
        self._test_graph_order({
            'base': [],
            'module3': ['module1'],
            'module4': ['module2', 'module3'],
            'module1': ['base'],
            'module5': ['module2', 'module4'],
            'module2': ['module1'],
        }, ['base', 'module1', 'module2', 'module3', 'module4', 'module5']
        )
```